src/Compiler/Checking/AttributeChecking.fs

```diff
@@ -24,8 +24,12 @@ open FSharp.Compiler.TypeProviders
 open FSharp.Core.CompilerServices
 #endif
 
-exception ObsoleteWarning of string * range
-exception ObsoleteError of string * range
+exception ObsoleteDiagnostic of
+    isError: bool *
+    diagnosticId: string *
```

Comment:
I wonder if diagId and urlFormat would be better represented as options, to have None there instead of an empty string when the value is missing. What do you think?

Reply:
I decided to keep this as `string` to be consistent with `message` with arguably can also be option tbh.

@psfinaki should we make all optional then ?

Reply:
Yeah I thought the message should be filled in, but I guess this is rather a guidance than a hard requirement, based on what I see in MSDN. So could also be an option...

Well this is just a suggestion to see if that would make things clearer or better. And well, I don't like primitive types in general, but that's a personal preference :D

Reply:
Yeah, if user don't provide a custom message. We will just show `This construct is deprecated`.  I don't mind changing this. Just say the word :)

Reply:
Would probably wait for another opinion so that you don't do double work... I'm not insisting here really :)